File: java/src/com/eliudarudo/cliapp/questionnaire/Utility.java
package com.eliudarudo.cliapp.questionnaire;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class Utility {

    public static class RandomGenerator {
        public RandomGenerator() {}

        public int randint(int min, int max) {
            Random random = new Random();
            return min + random.nextInt(max - min + 1);
        }
    }

    public static boolean stringIsValid(String string) {
        try {
            return string.length() != 0;
        } catch(Exception e) {
            return false;
        }
    }

    public static void printFieldErrorMessage() {
        System.out.println("Please try again, invalid input");
    }

    public static String fetchUserEvaluation(String defaultMessage, ArrayList<Question> correctQuestions) {
        if(correctQuestions == null) {
            return defaultMessage;
        }
        StringBuilder message = new StringBuilder();

        String frontend = "frontend";
        String backend = "backend";
        String devops = "devops";
        String mobile = "mobile";

        HashMap<String, Integer> developerRating = new HashMap();
        developerRating.put(frontend, 0);
        developerRating.put(backend, 0);
        developerRating.put(devops, 0);
        developerRating.put(mobile, 0);

        for(Question question: correctQuestions) {
            String tag = question.tag;
            developerRating.replace(tag, developerRating.get(tag) + 1);
        }

        int passingLimit = 3;
        int failureLimit = 2;

        boolean isFrontendDeveloper = developerRating.get(frontend) > failureLimit && developerRating.get(
                backend) <= failureLimit && developerRating.get(devops) <= failureLimit && developerRating.get(mobile) <= failureLimit;

        boolean isBackendDeveloper = developerRating.get(frontend) <= failureLimit && developerRating.get(
                backend) > failureLimit && developerRating.get(devops) <= failureLimit && developerRating.get(mobile) <= failureLimit;

        boolean isDevopsEngineer = developerRating.get(frontend) < failureLimit && developerRating.get(
                backend) < failureLimit && developerRating.get(devops) > failureLimit && developerRating.get(mobile) < failureLimit;

        boolean isMobileDeveloper = developerRating.get(frontend) < failureLimit && developerRating.get(
                backend) < failureLimit && developerRating.get(devops) < failureLimit && developerRating.get(mobile) > failureLimit;

        boolean isFullStackDeveloper = developerRating.get(frontend) > failureLimit && developerRating.get(
                backend) > failureLimit && developerRating.get(devops) < failureLimit;

        boolean isSoftwareEngineer = developerRating.get(frontend) > failureLimit && developerRating.get(
                backend) > failureLimit && developerRating.get(devops) >= failureLimit;

        boolean isAllRounder = developerRating.get(frontend) >= failureLimit && developerRating.get(
                backend) >= failureLimit && developerRating.get(devops) >= failureLimit && developerRating.get(mobile) >= failureLimit;
                
        // System.out.println("frontend developer: " + isFrontendDeveloper);
        // System.out.println("backend developer: " + isBackendDeveloper);
        // System.out.println("devops developer: " + isDevopsEngineer);
        // System.out.println("mobile developer: " + isMobileDeveloper);
        // System.out.println("fullstack developer: " + isFullStackDeveloper);
        // System.out.println("software engineer: " + isSoftwareEngineer);
        // System.out.println("all rounder: " + isAllRounder);

        if (isAllRounder) {
            message.append("WOW! Looks like we got an all rounder, congratulations!!!");
        }
        else if (isSoftwareEngineer) {
            message.append("You're definitely a Software engineer :-)");
        }
        else if (isFullStackDeveloper) {
            message.append("Congratulations, Oh Fullstack developer!");
        }
        else if (isDevopsEngineer) {
            message.append("You're definitely a DevOps engineer");
        }
        else if (isMobileDeveloper) {
            message.append("Looks like you are a mobile developer!");
        }
        else if (isBackendDeveloper) {
            message.append("Another great backend developer!");
        }
        else if (isFrontendDeveloper) {
            message.append("Looks like you are a frontend developer, right?");
        }
        else{
            message.append(defaultMessage);
        }

        message.append("\n");

        return message.toString();
    }
}

File: go/src/github.com/eliudarudo/go-cli-app/questionnaire/utility.go
package questionnaire

import (
	"fmt"
	"math/rand"
)

/* ---------------------------------------------- */

// Random struct
type Random struct{}

// RandInt struct
func (random *Random) RandInt(min int, max int) int {
	if max == 0 {
		return max
	}
	// rand.Seed(time.Now().UnixNano())

	return (rand.Intn(max-min) + min)
}

/* ---------------------------------------------- */

// StringIsValid function
func StringIsValid(str string) bool {
	if len(str) == 0 {
		return false
	}
	return true
}

// PrintFieldErrorMessage function
func PrintFieldErrorMessage() {
	fmt.Println("Please try again, invalid input")
}

// FetchUserEvaluation function
func FetchUserEvaluation(defaultMessage string, correctQuestions *[]Question) string {
	var message string

	developerRating := map[string]int{
		"frontend": 0,
		"backend":  0,
		"devops":   0,
		"mobile":   0,
	}

	for _, question := range *correctQuestions {
		tag := question.Tag
		developerRating[tag]++
	}

	frontend := "frontend"
	backend := "backend"
	devops := "devops"
	mobile := "mobile"

	// passingLimit := 3
	falureLimit := 2

	isFrontendDeveloper := developerRating[frontend] > falureLimit && developerRating[backend] <= falureLimit && developerRating[devops] <= falureLimit && developerRating[mobile] <= falureLimit

	isBackendDeveloper := developerRating[frontend] <= falureLimit && developerRating[backend] > falureLimit && developerRating[devops] <= falureLimit && developerRating[mobile] <= falureLimit

	isDevopsEngineer := developerRating[frontend] < falureLimit && developerRating[backend] < falureLimit && developerRating[devops] > falureLimit && developerRating[mobile] < falureLimit

	isMobileDeveloper := developerRating[frontend] < falureLimit && developerRating[backend] < falureLimit && developerRating[devops] < falureLimit && developerRating[mobile] > falureLimit

	isFullStackDeveloper := developerRating[frontend] > falureLimit && developerRating[backend] > falureLimit && developerRating[devops] < falureLimit

	isSoftwareEngineer := developerRating[frontend] > falureLimit && developerRating[backend] > falureLimit && developerRating[devops] >= falureLimit

	isAllRounder := developerRating[frontend] >= falureLimit && developerRating[backend] >= falureLimit && developerRating[devops] >= falureLimit && developerRating[mobile] >= falureLimit

	// fmt.Println("frontend developer: %v" , isFrontendDeveloper)
	// fmt.Println("backend developer: %v" , isBackendDeveloper)
	// fmt.Println("devops developer: %v" , isDevopsEngineer)
	// fmt.Println("mobile developer: %v" , isMobileDeveloper)
	// fmt.Println("fullstack developer: %v" , isFullStackDeveloper)
	// fmt.Println("software engineer: %v" , isSoftwareEngineer)
	// fmt.Println("all rounder: %v" , isAllRounder)

	if isAllRounder {
		message = "WOW! Looks like we got an all rounder, congratulations!!!"
	} else if isSoftwareEngineer {
		message = "You're definitely a Software engineer :-)"
	} else if isFullStackDeveloper {
		message = "Congratulations, Oh Fullstack developer!"
	} else if isDevopsEngineer {
		message = "You're definitely a DevOps engineer"
	} else if isMobileDeveloper {
		message = "Looks like you are a mobile developer!"
	} else if isBackendDeveloper {
		message = "Another great backend developer!"
	} else if isFrontendDeveloper {
		message = "Looks like you are a frontend developer, right?"
	} else {
		message = defaultMessage
	}

	message += "\n"

	return message
}
